arrows/vxl/pixel_feature_extractor.cxx

```diff
@@ -150,18 +190,75 @@ ::concatenate_images( std::vector< vil_image_view_base_sptr > filtered_images )
   return concatenated_planes;
 }
 
+// ----------------------------------------------------------------------------
+// Convert to a narrower type without wrapping
+template < typename out_t, typename in_t >
+vil_image_view< out_t >
+safe_narrowing_cast( vil_image_view< in_t > input_image )
+{
+  auto const ni = input_image.ni();
+  auto const nj = input_image.nj();
+  auto const np = input_image.nplanes();
+  vil_image_view< out_t > output_image{ ni, nj, np };
+
+  constexpr out_t max_output_value = std::numeric_limits< out_t >::max();
+  constexpr out_t min_output_value = std::numeric_limits< out_t >::min();
+
+  vil_transform(
+    input_image, output_image,
+    [=]( in_t pixel ){
+      if( pixel < min_output_value )
+      {
+        return min_output_value;
+      }
+      else if( pixel > max_output_value )
+      {
+        return max_output_value;
+      }
+      else
+      {
+        return static_cast< out_t >( pixel );
+      }
+    } );
+  return output_image;
+}
+
+// ----------------------------------------------------------------------------
+template < typename pix_t >
+vil_image_view< pix_t >
+convert_to_typed_vil_image_view(
+  kwiver::vital::image_container_sptr input_image,
+  bool input_has_larger_range = false )
```

Comment:
It looks like this is all just a to handle the special case of `variance`.  That's the only place we call this with `input_has_larger_range == true` right?   The more common operation when converting potentially unknown image types in VXL is to use [vil_convert_stretch_range](https://vxl.github.io/doc/release/core/vil/html/vil__convert_8h.html#a02186007bf72f35be646346d40038aea) rather than clipping the data.  However, I understand that this "variance" thing (which is only loosely related to variance) is supposed to be a cumulative thing that is capped at some upper value intentionally.  It's probably best just to more explicitly call `vil_clamp` on the variance case before casting to byte, since clamping is specific to the way you want to map this "variance" to bytes and is not the usual way you would want to map floating point images to byte images.

Reply:
Yes that's the case. I made it more explicit, but I still have a function which does both the clamping and casting.